### PR TITLE
Introduce binary Kryo serialization of Geometry types

### LINKED ISSUE #2581: Improve Geometry Kryo serialization
Starting with the following geometry:

Linestring

```scala
scala> KryoWrapper(linestring)
res0: geotrellis.spark.util.KryoWrapper[geotrellis.vector.Geometry] = geotrellis.spark.util.KryoWrapper@13f584a9

scala> res0.getValueSerialized
res1: Array[Byte] = Array(-80, 1, 1, 0, 0, 0, 0, 1, 0, 99, 111, 109, 46, 118, 105, 118, 105, 100, 115, 111, 108, 117, 116, 105, 111, 110, 115, 46, 106, 116, 115, 46, 103, 101, 111, 109, 46, 76, 105, 110, 101, 83, 116, 114, 105, 110, -25, 1, 0, 0, -111, 1, 1, 0, -109, 1, 1, -104, 1, 1, -103, 1, 1, 1, 70, 76, 79, 65, 84, 73, 78, -57, 0, 0, 0, 0, 0, 0, 0, 0, -110, 1, 1, -126, 1, 1, -61, 12, -113, 1, 1, 65, 82, -10, 73, -30, 60, 77, 30, 65, 83, 44, -10, -52, 56, -43, 28, 127, -8, 0, 0, 0, 0, 0, 0, -113, 1, 1, 65, 82, -10, 75, 86, 33, -73, -123, 65, 83, 45, 9, -120, 68, -25, -60, 127, -8, 0, 0, 0, 0, 0, 0, -113, 1, 1, 65, 82, -10, 83, -65, -38, 39, 21, 65, 83, 45, 23, -52, 102, 56, -73, 127, -8, 0, 0, 0, 0, 0, 0, -113, 1, 1, 65, 82, -10, 87, 53, 107, -60, -46, 65, 83, 45, 28, 28, -91, 85, 60...

scala> res1.size
res2: Int = 43353

scala> linestring.toWKB(0)
res4: Array[Byte] = Array(0, 0, 0, 0, 2, 0, 0, 6, 66, 65, 82, -10, 73, -30, 60, 77, 30, 65, 83, 44, -10, -52, 56, -43, 28, 65, 82, -10, 75, 86, 33, -73, -123, 65, 83, 45, 9, -120, 68, -25, -60, 65, 82, -10, 83, -65, -38, 39, 21, 65, 83, 45, 23, -52, 102, 56, -73, 65, 82, -10, 87, 53, 107, -60, -46, 65, 83, 45, 28, 28, -91, 85, 60, 65, 82, -10, 91, -89, -24, 92, -14, 65, 83, 45, 23, 102, 75, 94, -59, 65, 82, -10, 98, -81, -117, 7, -111, 65, 83, 45, 15, -13, 100, -24, -128, 65, 82, -10, 106, -13, -127, 3, -36, 65, 83, 45, 24, 31, -104, -116, 75, 65, 82, -10, 126, -75, 38, 48, -53, 65, 83, 45, 33, -26, 29, -36, -85, 65, 82, -10, -117, -78, -41, 91, 19, 65, 83, 45, 34, 63, -18, 120, 44, 65, 82, -10, -99, 46, 107, 73, -17, 65, 83, 45, 18, 17, 29, -88, -46, 65, 82, -10, -48, 50, 62, 95, -113,...

scala> res4.size
res5: Int = 25641
```

We can clearly improve performance of Geometry serialization.  And we should.

### PR DESCRIPTION
**Pull request by jpolchlo**
## Overview

We were serializing Geometry subclasses using the default Kryo serializer.  This produced fairly large buffers.  This PR introduces a custom serializer that encodes the geometry to a compressed WKB buffer (using Deflate compression).

Signed-off-by: jpolchlo <jpolchlopek@azavea.com>

Closes #2581 

### Checklist

- [x] `docs/CHANGELOG.rst` updated, if necessary

### Demo

The original version:
```scala
val geojson = scala.io.Source.fromFile("/tmp/natural-earth-boundaries/africa.geojson").mkString
MultiPolygon(geojson.parseGeoJson[JsonFeatureCollection].getAllMultiPolygons.flatMap(_.polygons))
KryoWrapper(res0).getValueSerialized.size
```
=> 103203

The new version yields 29858 for the same code block.

### CODE REVIEW

**Comment by jpolchlo** [comment]
It's not entirely clear to me if this will actually lower the amount of data shuffled by Spark.  If Spark compresses data prior to transmission, this will increase compute times due to a redundant compression step.

File: spark/src/main/scala/geotrellis/spark/io/kryo/GeometrySerializer.scala [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    val wkb = WKB.write(geom)
+
+    val (data, compressed) =
+      if (wkb.length > 1400) {
+        // Only bother deflating data larger than about 1 packet worth
+        val deflater = new Deflater
+        deflater.setInput(wkb)
+        deflater.finish
+        val baos = new ByteArrayOutputStream
+        val buffer = Array.ofDim[Byte](4096)
```
Comment (pomadchin) [inline spark/src/main/scala/geotrellis/spark/io/kryo/GeometrySerializer.scala]: Btw, we actually always can guess max possible compressed data length: initial data size + 10 header bytes (for compression level 0).

Comment (jpolchlo) [reply spark/src/main/scala/geotrellis/spark/io/kryo/GeometrySerializer.scala]: I don't think there's much value in exposing a default compression level option to the user.  This stuff should all be transparently more awesome to the user, and the default level balances speed and compression level.  I think the only argument to be made here is do we just put the dial to 11 and max out the compression level under the assumption that network transmission is so slow that the time savings for packet transmission more than make up for the time spent compressing to the highest degree?

**Comment by jpolchlo** [comment]
I added a config variable (`geotrellis.serialization.vector.compression.level`) to control this behavior.  Using `java.util.zip.Deflater.DEFAULT_COMPRESSION` by default.

**Comment by echeipesh** [comment]
@jpolchlo  I'm not sure it makes sense to compress the WKB. Spark has `spark.io.compression = true` as default which compresses the shuffle blocks before sending them over the network. `spark.io.compression.codec` defaults to `lz4` and used to be/can be `snappy`, both of which are faster than deflate.

If I recall the WKB bytes were already much smaller than what kryo comes up with? They should be better compressible to boot.

**Comment by jpolchlo** [comment]
Note: without the compression, we get less than 50% size reduction from the original.  No comment on size savings of the actual compressed shuffle blocks (i.e., no idea if this PR has a real-world effect).

**Comment by jpolchlo** [comment]
For the sake of completeness, the post-deflate compression of the original, non-binary serialized form of the sample geometry was 32808 as opposed to the deflated form of the binary representation at 29858 (about a 9% improvement).

**Comment by pomadchin** [comment]
Merging after travis would finish his work!

**Pull request merged**